CRDs should support watch of protobuf PartialObjectMetadata

smarterclayton opened the pull request:
Correctly ensure CRDs can be watched using protobuf when transformed to
PartialObjectMetadata. To do this we add a set of serializers allowed to
be used for "normal" requests (that return CRDs) while the serializers
supported by the infrastructure is broader and includes protobuf. During
negotatiation we check for transformation requests and protobuf is
excluded from non-transform requests.

As part of the change, correct an error message when the server returns
a 406 but the client doesn't accept the format to avoid confusing users
who set impossible Accept rules for CRDs (the dynamic client doesn't
support Protobuf, so if the server responds with a protobuf status the
message from the server is lost and the generic error was confusing).

Part of https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190322-server-side-get-to-ga.md

/kind bug

```release-note
NONE
```

smarterclayton committed:
CRDs should support watch of protobuf PartialObjectMetadata

Correctly ensure CRDs can be watched using protobuf when transformed to
PartialObjectMetadata. To do this we add a set of serializers allowed to
be used for "normal" requests (that return CRDs) while the serializers
supported by the infrastructure is broader and includes protobuf. During
negotatiation we check for transformation requests and protobuf is
excluded from non-transform requests.

As part of the change, correct an error message when the server returns
a 406 but the client doesn't accept the format to avoid confusing users
who set impossible Accept rules for CRDs (the dynamic client doesn't
support Protobuf, so if the server responds with a protobuf status the
message from the server is lost and the generic error was confusing).

smarterclayton commented:
Extracted from #76054, required for metadata client

smarterclayton commented:
@liggitt I should have pulled this out earlier

smarterclayton commented:
/retest

fedebongio commented:
/assign @jpbetz

yliaog commented:
/cc

smarterclayton commented:
/retest

smarterclayton commented:
Ping, still needs attention for 1.15

smarterclayton commented:
/retest

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/negotiation/negotiate.go:
```diff
@@ -225,7 +226,7 @@ func acceptMediaTypeOptions(params map[string]string, accepts *runtime.Serialize
 	}
 
-	if options.Convert != nil && !endpoint.AllowsConversion(*options.Convert, accepts.MediaTypeType, accepts.MediaTypeSubType) {
+	if !endpoint.AllowsConversion(options.Convert, accepts.MediaTypeType, accepts.MediaTypeSubType) {
```
this is pretty twisty, having `AllowsConversion` check a "no conversion" case

smarterclayton replied:
AllowsConversionAttribute.  Some now disallow non.

Another wording is AllowMediaType.

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/rest.go:
```diff
@@ -78,7 +83,21 @@ func (scope *RequestScope) err(err error, w http.ResponseWriter, req *http.Reque
 }
 
-func (scope *RequestScope) AllowsConversion(gvk schema.GroupVersionKind, mimeType, mimeSubType string) bool {
+func (scope *RequestScope) AllowsConversion(gvk *schema.GroupVersionKind, mimeType, mimeSubType string) bool {
```
this method is really misleading now... it's really hard to understand that gvk is nil for normal output, and only non-nil for special types

smarterclayton replied:
Is that hard?  Convert nil means “no conversion” is not terribly different from calls to scheme conversion

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go:
```diff
@@ -762,6 +774,16 @@ func (s unstructuredNegotiatedSerializer) SupportedMediaTypes() []runtime.Serial
 			EncodesAsText:    true,
 			Serializer:       json.NewYAMLSerializer(json.DefaultMetaFactory, s.creator, s.typer),
 		},
+		{
```
would it be clearer to put some sort of attribute on SerializerInfo indicating it can only be used for specific types, and then check in `acceptMediaTypeOptions` in the non-conversion case if the incoming `accepts` info allows an unconverted use?

smarterclayton replied:
Then we have to put two serializer infos with the same media type.  Didn’t want to do that today, because callers assume serializer info media type is unique.

dstrebel commented:
Hello! We are starting Code Freeze for 1.15. Is this PR still planned to be included in 1.15?

smarterclayton commented:
Yes

smarterclayton commented:
Renamed with the wording discussed on slack.  I added an additional test verifying an existing type is skipped (but we are already well covered there)

smarterclayton commented:
PTAL

liggitt commented:
/lgtm
/approve

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton committed:
Set integration tests to use distinct namespaces

TestWatchBasedManager was racing with the default namespace creation.
To fix that flake and to ensure integration tests using a shared etcd
don't accidentally overlap in the future, move the three main tests
using the default namespace to separate namespaces, and have
TestWatchBasedManager create that namespace before it runs.

Make StartTestServer wait for default namespace creation, which will
reduce other flakes until future changes completely remove use of default
namespace.

From a failed integration run:

	watch_manager_test.go:66: namespaces "default" not found
	watch_manager_test.go:66: namespaces "default" not found
	watch_manager_test.go:66: namespaces "default" not found

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
Updated with fix for integration flake (slightly worsened by this PR), PTAL

liggitt commented:
seen here too:
```
crd_test.go:64: Failed to create NetworkPolicy: namespaces "default" not found
```

how did this PR make the problem worse?

liggitt commented:
/lgtm

smarterclayton commented:
Typo in test case, retagging

smarterclayton commented:
/retest

smarterclayton commented:
/test pull-kubernetes-e2e-gce-big

smarterclayton commented:
/test pull-kubernetes-kubemark-e2e-gce-big
